### ISSUE TITLE
gem5 to Streamline apc file has error

### ISSUE THREAD

**Issue by atrah22** [gem5/gem5#145]
**Describe the bug**
util/streamline/m5stats2streamline.py script produced .apc file is invalid
--> in newer DS-5 streamline (29rel1), ultimate edition --> invalid apc format
--> in older DS-5 streamline (24rel1), ultimate edition --> error is,
Error detected in the binary data with respect to end-of-line markers. Check the settings on any FTP tools used and ensure any ASCII end-of-line conversions are disabled.
  java.io.IOException: Error detected in the binary data with respect to end-of-line markers. Check the settings on any FTP tools used and ensure any ASCII end-of-line conversions are disabled.
  	at com.arm.streamline.analysis.gator.GatorProtocolV17.processSummary(GatorProtocolV17.java:107)
  	at com.arm.streamline.analysis.gator.GatorProtocolV17.process(GatorProtocolV17.java:62)
  	at com.arm.streamline.analysis.gator.GatorDatabase.processOnce(GatorDatabase.java:127)
  	at com.arm.streamline.analysis.gator.GatorDatabase.processOnce(GatorDatabase.java:137)
  	at com.arm.streamline.analysis.gator.GatorDatabase.process(GatorDatabase.java:145)
  	at com.arm.streamline.analysis.AnalysisBase.run(AnalysisBase.java:59)
  	at java.lang.Thread.run(Thread.java:745)

**Affects version**
v22.1, v23.0

**gem5 Modifications**
The modification done mainly in the util/streamline/m5stats2streamline.py . 
1) in the function,

```

def string_to_int(string): #added function
  try:
    return int(string)
  except ValueError:
    return None

############################################################
# Write binary
############################################################
def writeBinary(outfile, binary_list):
    for i in binary_list: #added line
        if isinstance(i,str): #added line
            i_temp = string_to_int(i) #added line
            if (i_temp == None): #added line
                i = i #added line
            else: #added line
                i=i_temp #added line
        if isinstance(i,int): #added line
            i = chr(i) #added line
        byteVal = bytes(i, "utf-8") #added line
        #outfile.write(f"{i:c}") # original line
        outfile.write(byteVal) #added line
```

I see that in binary_list provided to writeBinary sometimes has combination of int and strings! . so just outfile.write(f"{i:c}") gives type error that byte like object need and not str.

2) packed32(x) function, where x has to be int and sometimes its  not int and its float!

```
def packed32(x):
    ret = []
    more = True
    while more:
        x =int(x) #added line
        b = x & 0x7F
        x = x >> 7
        if ((x == 0) and ((b & 0x40) == 0)) or (
            (x == -1) and ((b & 0x40) != 0)
        ):
            more = False
        else:
            b = b | 0x80
        ret.append(b)
    return ret


```

**Expected behavior**
Expected behavior is to see the waveform in arm DS-5 streamline for gem5 FS of arm-linux. I have the license for older streamline version that could be compatible with gem5 conversion script.

**Host Operating System**
Ubuntu 18.04

**Host ISA**
x864

**Reproduce**
I have enabled context switch dump. I attached the files,
```
python m5stats2streamline.py ./stat_config.ini ./streamline ./streamline.apc
zip -r ./streamline.apc.zip ./streamline.apc (then import the apc.zip file in DS-5 streamline)

```

**Comment by BobbyRBruce** [comment]
@giactra : You might know more about this? Though this code doesn't seem to have been updated since 2016. Maybe bit-rot? Should we remove this, maybe?

**Comment by giactra** [comment]
Thanks for reporting this issue. I have the feeling part of the problem is the stricter behaviour of python3 wrt to bytes/strings.

Have you tried running the script with python2.7? (Just out of curiosity, we should fix this anyway for python3)

I also didn't get if you managed to make the older streamline work with your modifications. If not I am wondering whether the APC file is actually expected to store the stringified version of an integer (in binary) rather than the integer itself...

**Comment by atrah22** [comment]
> 

Hello, thanks for the quick reply.
 
1) Yes, I tried with python 2.7 as well and got same error.
2) No, I couldn't make it work in the older version. The older streamline version I think, produces more detailed error log for the apc.zip provided. The newer version just reports invalid apc format.
3) I tried to convert as you mentioned, integer to string and convert to binary. Now the error shown in 5.24.0 "streamline is no events received from the target". I think, the first modification was close to the desired .apc format than this way. The changes for this try is shown below:

```
def writeBinary(outfile, binary_list):
    for i in binary_list:
        if isinstance(i,int):
            i = str(i)
        byteVal = bytes(i, "utf-8") #works for int
        #outfile.write(f"{i:c}")
        outfile.write(byteVal)

```
 Also, could there be change/update in .apc format of the DS-5 streamline?

**Comment by atrah22** [comment]
Hello, Update (Solution):
This below code worked. The idea is to save string "1" or "\n" etc as b"1" or b"\n"
and incase it is integer 1 ,  then in b'\x01' format.

```
def writeBinary(outfile, binary_list):
    for i in binary_list:
        if isinstance(i, str):
            byteVal = bytes(i, "utf-8")
        elif isinstance(i, int):
            byteVal = bytes([i])
        else:
            byteVal = i
            
        outfile.write(byteVal)
```

Thanks

**Comment by atrah22** [comment]
Hello, Update (solution):
I could now make it work.  Worked with streamline DS-5 v5.24 (Didn't work with v5.29)

```
def writeBinary(outfile, binary_list):
    for i in binary_list:
        if isinstance(i,int):
            byteVal = bytes([i]) 
        else if isinstance(i,str):
            byteVal = bytes(i, "utf-8") 
        else:
            byteVal = i
        
        outfile.write(byteVal)
```

When, its an integer, for e.g. 1 then its written as b'\x01'
When its a string, for e.g. "1"  or "\n" then its written as b'1' , b'\n'

Thanks,
Atul

**Comment by powerjg** [comment]
Thanks for follow up @giactra. I've assigned you to keep track of this issue. Don't feel that you have the responsibility to fix it, but we would appreciate you keeping track of the status of the issue and closing when appropriate :).

**Comment by giactra** [comment]
Thanks @atrah22 ! Do you want to create a PR to fix it? Otherwise I will create one myself

**Comment by atrah22** [comment]
> Thanks @atrah22 ! Do you want to create a PR to fix it? Otherwise I will create one myself

Hello @giactra , I created a PR (https://github.com/gem5/gem5/pull/207) thanks.

On a different note, any suggestions on a way to annotate my source code ROI (binary running in gem5) so that it can be visible in streamline ? This could be a feature request. I know, in case of real device (using gator) we can use streamline api to annotate ROI.

**Comment by BobbyRBruce** [comment]
Fixed by #211
